### PR TITLE
[Yaml] Simplify merging

### PR DESCRIPTION
**Pull request by GuilhemN**
| Q             | A
| ------------- | ---
| Branch?       | 3.4
| Bug fix?      | no
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | 
| License       | MIT
| Doc PR        | 

With this PR, merge nodes are like any other nodes and are managed only at the end of the block parse.
This permits to remove duplicated logic ~~but the resulting key order may change. According to the yaml spec, the order in mappings doesn't matter so I think it's not a big deal.~~

### CODE REVIEW

File: src/Symfony/Component/Yaml/Parser.php [hunk 1]
```diff
@@ -359,6 +299,31 @@ public function parse($value, $flags = 0)
...
+                    if (!is_array($mergedMapping)) {
+                        throw new ParseException('Merge items must be arrays.');
+                    }
+                }
+                $merged = call_user_func_array('array_merge', array_reverse($merged));
+            }
+
+            // If the value associated with the key is a single mapping node, each of its key/value pairs is inserted into the
+            // current mapping, unless the key already exists in it.
+            $data = array_merge($merged, $data);
```
Comment (GuilhemN) [inline src/Symfony/Component/Yaml/Parser.php]: I'm not sure that's a good idea because of numeric keys. Should we use the `+` operator?

**Comment by nicolas-grekas** [comment]
Ping @fabpot @Tobion since your wrote the current code. WDYT about the ordering change?

File: src/Symfony/Component/Yaml/Parser.php [hunk 2]
```diff
@@ -170,77 +169,18 @@ public function parse($value, $flags = 0)
...
 
-                if ($mergeNode) {
-                    // Merge keys
-                } elseif (!isset($values['value']) || '' == trim($values['value'], ' ') || 0 === strpos(ltrim($values['value'], ' '), '#')) {
+                if (!isset($values['value']) || '' == trim($values['value'], ' ') || 0 === strpos(ltrim($values['value'], ' '), '#')) {
                     // hash
                     // if next line is less indented or equal, then it means that the current value is null
                     if (!$this->isNextLineIndented() && !$this->isNextLineUnIndentedCollection()) {
                         // Spec: Keys MUST be unique; first one wins.
                         // But overwriting is allowed when a merge node is used in current block.
```
Comment (xabbuh) [inline src/Symfony/Component/Yaml/Parser.php]: As we do not handle merges anymore here I would remove this comment.

File: src/Symfony/Component/Yaml/Parser.php [hunk 3]
```diff
@@ -359,6 +299,31 @@ public function parse($value, $flags = 0)
             mb_internal_encoding($mbEncoding);
         }
 
+        if (array_key_exists('<<', $data)) {
```
Comment (xabbuh) [inline src/Symfony/Component/Yaml/Parser.php]: Are you sure that this still works if there are several occurrences of `<<` in one YAML document?

File: src/Symfony/Component/Yaml/Tests/Fixtures/sfMergeKey.yml [hunk 4]
```diff
@@ -43,11 +43,11 @@ yaml: |
 php: |
     array(
         'foo' => array('a' => 'Steve', 'b' => 'Clark', 'c' => 'Brian'),
-        'bar' => array('a' => 'before', 'd' => 'other', 'b' => 'new', 'c' => array('foo' => 'bar', 'bar' => 'foo'), 'x' => 'Oren'),
+        'bar' => array('a' => 'before', 'b' => 'new', 'c' => array('foo' => 'bar', 'bar' => 'foo'), 'd' => 'other', 'x' => 'Oren'),
```
Comment (xabbuh) [inline src/Symfony/Component/Yaml/Tests/Fixtures/sfMergeKey.yml]: What is the reason to change the fixture file? Is it because of the changed order of keys in the result you talked about in the issue description? If so, I suggest that we look for how to adjust the test instead. If there is not a solution, I would not make this change as it could mean that people rely on this too and we would break the applications then.

File: src/Symfony/Component/Yaml/Parser.php (hunk 3 above)
Comment (GuilhemN) [inline src/Symfony/Component/Yaml/Parser.php]: Good catch, fixed and deprecated at the same time.

Comment (xabbuh) [inline src/Symfony/Component/Yaml/Parser.php]: Are you sure that we must deprecate this?

Comment (GuilhemN) [inline src/Symfony/Component/Yaml/Parser.php]: It is not supported by the spec, it is built [on top of it](http://yaml.org/type/merge.html), so imo it should work like any other nodes and multiple occurrences should not be supported.

Comment (xabbuh) [inline src/Symfony/Component/Yaml/Parser.php]: Indeed, that's a valid point. However, I think we should do that in a different PR (and this should then also be documented in the upgrade files).

File: src/Symfony/Component/Yaml/Tests/Fixtures/sfMergeKey.yml (hunk 4 above)
Comment (GuilhemN) [inline src/Symfony/Component/Yaml/Tests/Fixtures/sfMergeKey.yml]: This order change is only happening when the merge node is not at the beginning of the mapping. I really think people should not rely on that behavior but I can fix it when using only one node (which should always be the case).

File: src/Symfony/Component/Yaml/Parser.php (hunk 3 above)
Comment (GuilhemN) [inline src/Symfony/Component/Yaml/Parser.php]: ok, updated

**Comment by GuilhemN** [comment]
I was able to keep the order the same as now when using only one merge node; it may change however if using several but again, I don't think that's a big deal :)

File: src/Symfony/Component/Yaml/Parser.php (hunk 3 above)
Comment (xabbuh) [inline src/Symfony/Component/Yaml/Parser.php]: `isset()` should be enough now, right?

Comment (GuilhemN) [inline src/Symfony/Component/Yaml/Parser.php]: We should be able to detect if someone uses `null` and throw a proper exception.

Comment (xabbuh) [inline src/Symfony/Component/Yaml/Parser.php]: But above we always ensure that we add an array for the `<<` key. So if we want to throw an exception for `null`, we must do it there.

Comment (GuilhemN) [inline src/Symfony/Component/Yaml/Parser.php]: We ensure that only when there are multiple merge nodes, this check will still be useful for the following example:
```yml
foo:
    <<: ~
```

File: src/Symfony/Component/Yaml/Parser.php [hunk 5]
```diff
@@ -250,19 +189,29 @@ public function parse($value, $flags = 0)
...
                     // Spec: Keys MUST be unique; first one wins.
-                    // But overwriting is allowed when a merge node is used in current block.
-                    if ($allowOverwrite || !isset($data[$key])) {
+                    if (!isset($data[$key])) {
                         $data[$key] = $value;
+                    } elseif ('<<' === $key) {
+                        if (isset($data['<<'][0])) {
+                            $data['<<'][] = $value;
+                        } else {
+                            $data['<<'] = array($data['<<'], $value);
```
Comment (xabbuh) [inline src/Symfony/Component/Yaml/Parser.php]: This will break the case where `$data['<<']` already is an array (see https://3v4l.org/9Tb7d).

File: src/Symfony/Component/Yaml/Parser.php (hunk 3 above)
Comment (xabbuh) [inline src/Symfony/Component/Yaml/Parser.php]: Actually, that example should throw an exception. I would trigger it above when the scalar value would be set.

File: src/Symfony/Component/Yaml/Parser.php [hunk 6]
```diff
@@ -359,6 +308,47 @@ public function parse($value, $flags = 0)
             mb_internal_encoding($mbEncoding);
         }
 
+        if (array_key_exists('<<', $data)) {
+            $merged = $data['<<'];
```
Comment (xabbuh) [inline src/Symfony/Component/Yaml/Parser.php]: In case of multiple merge keys, this would be an array of arrays and we will need to traverse it. Always ensuring that `$data['<<']` is an array would make the traversing easier and would allow to use `isset()`.

**Comment by sstok** [comment]
I'm not sure I fully understand, what do you mean with "but the resulting key order may change"?

I know the Config component in some cases relies on the order of defining, the firewall configuration for example. So would this change now?

File: src/Symfony/Component/Yaml/Parser.php (hunk 5 above)
Comment (GuilhemN) [inline src/Symfony/Component/Yaml/Parser.php]: That's why there is the first case :)

File: src/Symfony/Component/Yaml/Parser.php (hunk 6 above)
Comment (GuilhemN) [inline src/Symfony/Component/Yaml/Parser.php]: I'm not sure this is worth it, the check would have to be added in two places while it is only in one now.

File: src/Symfony/Component/Yaml/Parser.php (hunk 3 above)
Comment (GuilhemN) [inline src/Symfony/Component/Yaml/Parser.php]: It throws an exception but later.

**Comment by GuilhemN** [comment]
@sstok sequences order is kept, this only affects mappings with multiple merge nodes (more than one).
The order of keys may change in this case (for example, `{foo: bar, quz: one}` becomes `{quz: one, foo: bar}`) but it is really an edge case and I don't think it will affect anyone.

**Comment by GuilhemN** [comment]
rebased

**Comment by xabbuh** [comment]
Not sure if making this change in `master` is a good idea. Imagine we fix a bug related to the merge handling in 2.7. Merging branches up will then become a nightmare due to these changes.

**Comment by GuilhemN** [comment]
Should I rebase this on 2.7 then?

**Comment by Tobion** [comment]
I don't see the simplification in this PR and it's also not less code (´+59 −56`).

**Comment by GuilhemN** [comment]
@Tobion everytime you update [these lines](https://github.com/GuilhemN/symfony/blob/1ec1bf86480a041598af04de8b4fa56e91e968bd/src/Symfony/Component/Yaml/Parser.php#L207-L255), most of the times you also have to update the lines currently managing merge nodes (which are removed in this PR); with this change you won't have to do it twice, so yes imo it's a simplification.

> it's also not less code (´+59 −56`).

That's only because we allow multiple merge nodes (should be deprecated and removed in 4.0 imo) and because the current design is meant to keep the keys order that was given before. Nothing that can't be fixed to simplify this even more in 4.0.

**Comment by GuilhemN** [comment]
I don't have the time to rebase this PR for now so I'm closing it, maybe I'll reopen it later.
